### ISSUE TITLE
Problem with queries which contain blank strings

### ISSUE THREAD

**Issue by M1ke** [auraphp/Aura.Sql#107]
Using a query of the following format

```
UPDATE table SET `value`=:value, `blank`='', `value2` = :value2, `blank2` = '', `value3`=:value3 WHERE id = :id
```

And passing in an array formatted as:

```
$arr = ['value'=> 'string', 'id'=> 1, 'value2'=> 'string', 'value3'=>'string'];
```

I receive the PDO " number of bound variables does not match number of tokens" error.

Examining the situation the `Rebuilder` returns a valid query (i.e. the query that went in) but the array has been truncated to:

```
Array
(
    [value] => string
    [value3] => string
    [id] => 1
)
```

This happens because the `preg_split` in `rebuildStatement` is splitting these "empty string" declarations incorrectly. The more empty strings in the query the more the problems. I am uncertain if the order of the strings matters.

I have added a temporary fix in my fork which adds:

```
$statement = str_replace("''", "_NULL_", $statement);
...
$statement = str_replace("_NULL_", "''", $statement);
```

Either side of the `list($statement, $values) = $rebuilder->__invoke($statement, $values);` line. This could also be added inside the Rebuilder.

Is this a legitimate problem or is there a better way to be writing the original query such that ExtendedPDO doesn't support the method in question?

**Comment by pmjones** [comment]
From what I remember about the rebuilder, it counts the number of quotes, and looks for closing sets of quotes. That might be the culprit here.

A workaround would be to use placeholders for the blank values ...

```
UPDATE table SET `value`=:value, `blank`= :blank, `value2` = :value2, `blank2` = :blank2, `value3`=:value3 WHERE id = :id
```

... and then to pass blank `''` values in the array for those placeholders.

That's a workaround for the interim.

The `str_replace()` solution presumes that there are no occurrences of `_NULL_` in other literals in the SQL string, which may not be true; in those cases, the query might be modified in unexpected ways. Probably not something that can be added to the library as-is.

However, if you can submit a failing test in a PR, I or someone else can eventually get around to working on the regex.

How does all that sound?

**Comment by harikt** [comment]
I was also thinking to write what @pmjones suggested passing placeholders.

**Comment by M1ke** [comment]
Are there any tests for the regex at present? The issue is that I can't be certain what it's meant to detect or not detect as I don't understand the whole syntax at present, or what the goal is.

**Comment by harikt** [comment]
@M1ke I think there is no tests for regular expression itself.  Don't know / remember what was the decision to extract it. But it happened in 2.1.0. iirc it was something with the place holder issues I guess.

**Comment by nyamsprod** [comment]
Another workaround is to change your request and to have a newline between conditions:

``` php
UPDATE table SET 
    `value`=:value, 
    `blank`='',
    `value2` = :value2,
    `blank2` = '',
    `value3`=:value3
WHERE id = :id
```

This will even have the side effect of rendering you SQL query more readable. We bumped into this bug too at work and that's the workaround we choose to implement

**Comment by harikt** [comment]
@nyamsprod great! .

Thank you for the patch.

**Comment by frederikbosch** [comment]
Is this bug not also related to #104 and #108? In other words: when merging 108, this should be fixed, right?

**Comment by pmjones** [comment]
Note to self: add a test for this against #134 after it merges.

**Comment by pmjones** [comment]
#134 is obviated; write a test for this against #138 .

**Comment by pmjones** [comment]
This appears to be fixed by #138. Thanks for the report!
